DOCSP-41972: GridFS guide (#133)
mongodb/docs-php-library: source/write/gridfs.txt

size limit of 16MB. For more detailed information on whether GridFS is
suitable for your use case, see :manual:`GridFS </core/gridfs>` in the
{+mdb-server+} manual.

How GridFS Works
----------------

GridFS organizes files in a **bucket**, a group of MongoDB collections
that contain the chunks of files and information describing them. The
bucket contains the following collections, named using the convention
defined in the GridFS specification:

- The ``chunks`` collection stores the binary file chunks.
- The ``files`` collection stores the file metadata.

When you create a new GridFS bucket, the library creates the preceding 
collections, prefixed with the default bucket name ``fs``, unless
you specify a different name. The library also creates an index on each
collection to ensure efficient retrieval of the files and related
metadata. The library creates the GridFS bucket, if it doesn't exist, only when the first write
operation is performed. The library creates indexes only if they don't exist and when the
bucket is empty. For more information about
GridFS indexes, see :manual:`GridFS Indexes </core/gridfs/#gridfs-indexes>`
in the {+mdb-server+} manual.

> When storing files with GridFS, the library splits the files into smaller

Reviewer: ```suggestion
When using GridFS to store files, the library splits the files into smaller
```

chunks, each represented by a separate document in the ``chunks`` collection.
It also creates a document in the ``files`` collection that contains
a file ID, file name, and other file metadata. You can upload the file from
memory or from a stream. See the following diagram to see how GridFS splits

> the files when uploaded to a bucket.

Reviewer: ```suggestion
the files when uploaded to a bucket:
```

bucket contains the following collections, named using the convention
defined in the GridFS specification:

- The ``chunks`` collection stores the binary file chunks.
- The ``files`` collection stores the file metadata.

When you create a new GridFS bucket, the library creates the preceding 
collections, prefixed with the default bucket name ``fs``, unless
you specify a different name. The library also creates an index on each
collection to ensure efficient retrieval of the files and related
metadata. The library creates the GridFS bucket, if it doesn't exist, only when the first write
operation is performed. The library creates indexes only if they don't exist and when the
bucket is empty. For more information about
GridFS indexes, see :manual:`GridFS Indexes </core/gridfs/#gridfs-indexes>`
in the {+mdb-server+} manual.

When using GridFS to store files, the library splits the files into smaller
chunks, each represented by a separate document in the ``chunks`` collection.
It also creates a document in the ``files`` collection that contains
a file ID, file name, and other file metadata. You can upload the file from

> memory or from a stream. View the following diagram to see how GridFS splits

Author: Updated!

method on your database. This method accesses an existing bucket or creates
a new bucket if one does not exist.

The following example calls the ``selectGridFSBucket()`` method on the ``db``
database:

.. literalinclude:: /includes/write/gridfs.php
   :language: php
   :dedent:
   :start-after: start create bucket
   :end-before: end create bucket

You can also explicitly create a GridFS bucket by calling the
constructor for the ``MongoDB\GridFS\Bucket`` class and passing the following
parameters:

- ``MongoDB\Driver\Manager`` object, which maintains connections between the
  PHP extension and MongoDB
- Name of the database, as a string, for which to create the bucket

The following example explicitly constructs a GridFS bucket:

.. literalinclude:: /includes/write/gridfs.php
   :language: php
   :dedent:
   :start-after: start create bucket explicit
   :end-before: end create bucket explicit

.. _gridfs-create-custom-bucket:

Customize the Bucket Name
~~~~~~~~~~~~~~~~~~~~~~~~~

To create or reference a bucket with a custom name other than the default name
``fs``, pass an options array to the ``selectGridFSBucket()`` method that sets

> the ``bucketName`` option. 

Author: Added an options table here

The following example creates a bucket named ``'myCustomBucket'``:

.. literalinclude:: /includes/write/gridfs.php
   :language: php
   :dedent:
   :start-after: start create custom bucket
   :end-before: end create custom bucket

.. _gridfs-upload-files:

> Upload Files

Reviewer 2: It may be helpful to also demonstrate `uploadFromStream()` from the second code examples in [Uploading Files with Writable Streams](https://www.mongodb.com/docs/php-library/current/tutorial/gridfs/#uploading-files-with-writable-streams). That's the suggested API for copying a local file into GridFS.

the {+php-library+}, see the :ref:`<php-cursors>` guide.

The following code example shows you how to retrieve and print file metadata
from files in a GridFS bucket. It uses a ``foreach`` loop to iterate through
the returned cursor and display the contents of the file uploaded in the
:ref:`gridfs-upload-files` example:

.. io-code-block::
   :copyable:

   .. input:: /includes/write/gridfs.php
      :start-after: start retrieve file info
      :end-before: end retrieve file info
      :language: php
      :dedent:

   .. output::
      :visible: false

      {"_id":{"$oid":"..."},"chunkSize":1048576,"filename":"my_file",
      "length":13,"uploadDate":{"$date":{"$numberLong":"..."}},"metadata":
      {"contentType":"text\/plain"},"md5":"6b24249b03ea3dd176c5a04f037a658c"}

The ``find()`` method accepts various query specifications. You can use its
``$options`` parameter to specify the sort order, maximum number of documents to return,
and the number of documents to skip before returning. To view a list of available
options, see the `API documentation <{+api+}/method/MongoDBGridFSBucket-find/#parameters>`__.

.. _gridfs-download-files:

> Download Files

Reviewer 2: Consider `downloadToStream()`, from the second code example in [Downloading Files with Readable Streams](https://www.mongodb.com/docs/php-library/current/tutorial/gridfs/#downloading-files-with-readable-streams).

If we think the use case of downloading from GridFS to a file is common enough, that's the most convenient and efficient way to do so.

   stream needs to pull a chunk from MongoDB, it pulls the entire chunk
   into memory. The 255-kilobyte default chunk size is usually
   sufficient, but you can reduce the chunk size to reduce memory
   overhead.

.. _gridfs-rename-files:

Rename Files
------------

Use the ``MongoDB\GridFS\Bucket::rename()`` method to update the name of
a GridFS file in your bucket. You must specify the file to rename by its
``_id`` field rather than its file name.

The following example shows how to update the ``filename`` field to
``'new_file_name'`` by referencing a document's ``_id`` field:

.. literalinclude:: /includes/write/gridfs.php
   :language: php
   :dedent:
   :start-after: start rename files
   :end-before: end rename files

.. note::

   The ``rename()`` method supports updating the name of only one file at
   a time. To rename multiple files, retrieve a list of files matching the
   file name from the bucket, extract the ``_id`` field from the files you
   want to rename, and pass each value in separate calls to the ``rename()``
   method.

.. _gridfs-delete-files:

Delete Files
------------

Use the ``MongoDB\GridFS\Bucket::delete()`` method to remove a file's collection

> document and associated chunks from your bucket. This effectively deletes the file.

Reviewer 2: May be worth clarifying that this is going to delete a _revision_ of a file. If another `fs.files` document exists for that filename, it can still be read. And if users aren't aware of GridFS' revision behavior and selecting the most recent by default, they could be surprised to discover that they can still query for a filename after calling `delete()`. The clarification should help avoid that.

Note that this also applies to `rename()`. Common advice for both may be telling users to collect all IDs for a filename if they intend to rename or delete all of its revisions. I don't think you need example code for that (can be left as an exercise for the reader).

Author: I modified the note below the code examples with this information

Reviewer 2: I see the new "File Revisions" notes for renames/deletes. Is there a DOCSP ticket to track future work of documenting the `revision` option for download methods? That seems like it'd be generally relevant to all driver GridFS tutorials.

Author: Not currently, although a few pages have a section on revisions (like [Java](https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/gridfs/#file-revisions)) or mention it somewhere on the page

You must specify the file by its ``_id`` field rather than its file name.

The following example shows you how to delete a file by referencing its ``_id`` field:

.. literalinclude:: /includes/write/gridfs.php
   :language: php
   :dedent:
   :start-after: start delete files
   :end-before: end delete files

.. note::

   The ``delete()`` method supports deleting only one file at a time. To
   delete multiple files, retrieve the files from the bucket, extract
   the ``_id`` field from the files you want to delete, and pass each value
   in separate calls to the ``delete()`` method.

API Documentation
-----------------

To learn more about using the {+php-library+} to store and retrieve large files,
see the following API documentation:

- :phpmethod:`MongoDB\Database::selectGridFSBucket()`

> - :phpmethod:`MongoDB\GridFS\Bucket::__construct()`

Reviewer 2: This can be removed.

file name. The ``openUploadStream()`` method allows you to specify configuration
information in an options array, which you can pass as a parameter.

This example uses an upload stream to perform the following
actions:

- Opens a writable stream for a new GridFS file named ``'my_file'``
- Sets the ``metadata`` option in an array parameter
  to the ``openUploadStream()`` method
- Calls the ``fwrite()`` method to write data to ``'my_file'``, which the stream points to
- Calls the ``fclose()`` method to close the stream pointing to ``'my_file'`` 

.. literalinclude:: /includes/write/gridfs.php
   :language: php
   :dedent:
   :start-after: start-open-upload-stream
   :end-before: end-open-upload-stream

Upload an Existing Stream
~~~~~~~~~~~~~~~~~~~~~~~~~

Use the ``uploadFromStream()`` method to upload the contents of a stream to
a new GridFS file. The ``uploadFromStream()`` method allows you to specify configuration
information in an options array, which you can pass as a parameter.

This example performs the following actions:

- Calls the ``fopen()`` method to create a temporary memory-based stream
  that you can write to and read from

> - Calls the ``fwrite()`` method to write data to the stream

Reviewer 2: If you take my suggestion on the example code, these can be reduced to a single bullet that calls `fopen()` on a local file. You can also note that we open the file in "binary read" mode (i.e. `rb`).

For the download example, we would `fopen()` or destination in "binary write" mode (i.e. `wb`).

- Prints the file contents
- Calls the ``fclose()`` method to close the download stream pointing to ``'my_file'`` 

.. io-code-block::
   :copyable:

   .. input:: /includes/write/gridfs.php
      :start-after: start-open-download-stream-name
      :end-before: end-open-download-stream-name
      :language: php
      :dedent:

   .. output::
      :visible: false

      "Data to store"

.. note::

   If there are multiple documents with the same file name,
   GridFS will stream the most recent file with the given name (as
   determined by the ``uploadDate`` field).

Alternatively, you can use the ``MongoDB\GridFS\Bucket::openDownloadStream()``
method, which takes the ``_id`` field of a file as a parameter:

.. literalinclude:: /includes/write/gridfs.php
   :language: php
   :dedent:
   :start-after: start-download-files-id
   :end-before: end-download-files-id

.. note::

   The GridFS streaming API cannot load partial chunks. When a download
   stream needs to pull a chunk from MongoDB, it pulls the entire chunk
   into memory. The 255-kilobyte default chunk size is usually
   sufficient, but you can reduce the chunk size to reduce memory

>    overhead.

Reviewer 2: Is this a suitable place to talk about _increasing_ the chunk size if users will frequently work with large GridFS files?

This looks like it might have been copied from somewhere, so feel free to disregard. I'm not sure if this tutorial is the appropriate place to discuss that, but I think it's more common that users would want to increase the default chunk size than reduce it further.

Author: Added a note about increasing the chunk size

Download to an Existing Stream
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can download the contents of a GridFS file to an existing stream
by calling the ``MongoDB\GridFS\Bucket::downloadToStream()`` method
on your bucket.

This example performs the following actions:

> - Calls the ``fopen()`` method to open a download stream

Reviewer 2: Flip side of my earlier common on `uploadFromStream()`, this could be changed to `fopen()` the an output file using "binary write" mode.

   :language: php
   :dedent:
   :start-after: start-download-files-id
   :end-before: end-download-files-id

.. note::

   The GridFS streaming API cannot load partial chunks. When a download
   stream needs to pull a chunk from MongoDB, it pulls the entire chunk
   into memory. The 255-kilobyte default chunk size is usually
   sufficient, but you can reduce the chunk size to reduce memory
   overhead or increase the chunk size when working with larger files.
   For more information about setting the chunk size, see the 
   :ref:`gridfs-create-custom-bucket` section of this page.

Download a File Revision
~~~~~~~~~~~~~~~~~~~~~~~~

When your bucket contains multiple files that share the same file name,
GridFS chooses the most recently uploaded version of the file by default.
To differentiate between each file that shares the same name, GridFS assigns them
a revision number, ordered by upload time.

The original file revision number is ``0`` and the next most recent file revision
number is ``1``. You can also specify negative values that correspond to the recency
of the revision. The revision value ``-1`` references the most recent revision and ``-2``

> references the next most recent revision.

Reviewer 2: Excellent summary :+1:

You can instruct GridFS to download a specific file revision by passing an options
array to the ``openDownloadStreamByName()`` method and specifying the ``revision``
option. The following example reads the contents of the original file named
``'my_file'`` rather than the most recent revision:

.. literalinclude:: /includes/write/gridfs.php
   :language: php
   :dedent:
   :start-after: start-download-file-revision
   :end-before: end-download-file-revision

Download to an Existing Stream
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can download the contents of a GridFS file to an existing stream
by calling the ``MongoDB\GridFS\Bucket::downloadToStream()`` method
on your bucket.

This example performs the following actions:

- Calls the ``fopen()`` method to open a file located at ``/path/to/output_file``
  as a stream in binary write (``wb``) mode

> - Downloads a GridFS file that has an ``_id`` value of ``ObjectID('66e0a5487c880f844c0a32b1')``

Reviewer 2: ```suggestion
- Downloads a GridFS file that has an ``_id`` value of ``ObjectId('66e0a5487c880f844c0a32b1')``
```
